lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -0,0 +1,925 @@
+    name to differentiate between "incoming" and "outgoing" remote write interactions is necessary.
+    """
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        endpoint_schema: str = "http",
+        endpoint_address: Optional[str] = None,
+        endpoint_port: Union[str, int] = 9090,
+        endpoint_path: str = "/api/v1/write",
+    ):
+        """API to manage a provided relation with the `prometheus_remote_write` interface.
+
+        Args:
+            charm: The charm object that instantiated this class.
+            relation_name: Name of the relation with the `prometheus_remote_write` interface as
+                defined in metadata.yaml.
+            endpoint_schema: The URL schema for your remote_write endpoint. Defaults to `http`.
+            endpoint_address: The URL host for your remote_write endpoint as reachable
+                from the client. This might be either the pod IP, or you might want to
+                expose an address routable from outside the Kubernetes cluster, e.g., the
+                host address of an Ingress. If not provided, it defaults to the relation's
+                `bind_address`.
+            endpoint_port: The URL port for your remote_write endpoint. Defaults to `9090`.
+            endpoint_path: The URL path for your remote_write endpoint.
+                Defaults to `/api/v1/write`.
+
+        Raises:
+            RelationNotFoundError: If there is no relation in the charm's metadata.yaml
+                with the same name as provided via `relation_name` argument.
+            RelationInterfaceMismatchError: The relation with the same name as provided
+                via `relation_name` argument does not have the `prometheus_scrape` relation
+                interface.
+            RelationRoleMismatchError: If the relation with the same name as provided
+                via `relation_name` argument does not have the `RelationRole.requires`
+                role.
+        """
+        _validate_relation_by_interface_and_direction(
+            charm, relation_name, RELATION_INTERFACE_NAME, RelationRole.provides
+        )
+
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self._endpoint_schema = endpoint_schema
+        self._endpoint_address = endpoint_address
+        self._endpoint_port = int(endpoint_port)
+        self._endpoint_path = endpoint_path
+
+        on_relation = self._charm.on[self._relation_name]
+        self.framework.observe(
+            on_relation.relation_created,
+            self._set_endpoint_on_relation_change,
+        )
+        self.framework.observe(
+            on_relation.relation_joined,
+            self._set_endpoint_on_relation_change,
+        )
+
+    def update_endpoint(self, relation: Optional[Relation] = None) -> None:
```

Comment:
`Optional` is inferred when `None` is assigned
```suggestion
    def update_endpoint(self, relation: Relation = None) -> None:
```